### PR TITLE
Differentiation Utils

### PR DESCRIPTION
**Pull request by GreatRSingh**
## Description

Fix #(issue)

This contains Editable Module (From XiTorch) which is heavily used by DFT.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/utils/attribute_utils.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+import re
+import ast
+
+__all__ = ["get_attr", "set_attr", "del_attr"]
+
+# pattern to split the names, e.g. "model.params[1]" into ["model", "params", "[1]"]
+sp = re.compile(r"\[{0,1}[\"']{0,1}\w+[\"']{0,1}\]{0,1}")
+
+
+def get_attr(obj, name):
```
Comment (rbharath) [inline deepchem/utils/attribute_utils.py:14]: Add type annotations to this and other functions

File: deepchem/utils/attribute_utils.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    Returns
+    -------
+    val : object
+        The value of the attribute.
+
+    """
+    return _get_attr(obj, _preproc_name(name))
+
+
+def set_attr(obj, name, val):
```
Comment (rbharath) [inline deepchem/utils/attribute_utils.py:43]: Type annotations

File: deepchem/utils/attribute_utils.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        The value to set the attribute to.
+
+    Returns
+    -------
+
+    """
+    return _set_attr(obj, _preproc_name(name), val)
+
+
+def del_attr(obj, name):
```
Comment (rbharath) [inline deepchem/utils/attribute_utils.py:77]: Type annotations

File: deepchem/utils/attribute_utils.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    ...     obj.a
+    ... except AttributeError:
+    ...     print("AttributeError")
+    AttributeError
+
+    """
+    return _del_attr(obj, _preproc_name(name))
+
+
+def _get_attr(obj, names):
```
Comment (rbharath) [inline deepchem/utils/attribute_utils.py:99]: Add a docstring and type annotations

File: deepchem/utils/attribute_utils.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    attrfcn = lambda obj, name: getattr(obj, name)  # noqa: E731
+    dictfcn = lambda obj, key: obj.__getitem__(key)  # noqa: E731
+    listfcn = lambda obj, key: obj.__getitem__(key)  # noqa: E731
+    return _traverse_attr(obj, names, attrfcn, dictfcn, listfcn)
+
+
+def _set_attr(obj, names, val):
```
Comment (rbharath) [inline deepchem/utils/attribute_utils.py:106]: Add a docstring and type annotations

File: deepchem/utils/attribute_utils.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    attrfcn = lambda obj, name: setattr(obj, name, val)  # noqa: E731
+    dictfcn = lambda obj, key: obj.__setitem__(key, val)  # noqa: E731
+    listfcn = lambda obj, key: obj.__setitem__(key, val)  # noqa: E731
+    return _traverse_attr(obj, names, attrfcn, dictfcn, listfcn)
+
+
+def _del_attr(obj, names):
```
Comment (rbharath) [inline deepchem/utils/attribute_utils.py:113]: Add a docstring and type annotations

File: deepchem/utils/attribute_utils.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    dictfcn = lambda obj, key: obj.__delitem__(key)  # noqa: E731
+
+    def listfcn(obj, key):
+        obj.__delitem__(key)
+        obj.insert(key, None)  # to preserve the length
+
+    return _traverse_attr(obj, names, attrfcn, dictfcn, listfcn)
+
+
+def _preproc_name(name):
```
Comment (rbharath) [inline deepchem/utils/attribute_utils.py:124]: Add a docstring and type annotations

File: deepchem/utils/attribute_utils.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    return sp.findall(name)
+
+
+def _traverse_attr(obj, names, attrfcn, dictfcn, listfcn):
```
Comment (rbharath) [inline deepchem/utils/attribute_utils.py:128]: Add a docstring and type annotations

File: deepchem/utils/attribute_utils.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    if len(names) == 1:
+        return _applyfcn(obj, names[0], attrfcn, dictfcn, listfcn)
+    else:
+        return _applyfcn(_get_attr(obj, names[:-1]), names[-1], attrfcn,
+                         dictfcn, listfcn)
+
+
+def _applyfcn(obj, name, attrfcn, dictfcn, listfcn):
```
Comment (rbharath) [inline deepchem/utils/attribute_utils.py:136]: Add a docstring and type annotations

File: deepchem/utils/differentiation_utils/editable_module.py [hunk 10]
```diff
@@ -0,0 +1,502 @@
...
+from deepchem.utils.attribute_utils import get_attr, set_attr, del_attr
+
+__all__ = ["EditableModule"]
+
+torch_float_type = [torch.float32, torch.float, torch.float64, torch.float16]
+
+
+class EditableModule(object):
+    """
+    ``EditableModule`` is a base class to enable classes that it inherits be
```
Comment (rbharath) [inline deepchem/utils/differentiation_utils/editable_module.py:20]: Add a usage example

File: deepchem/utils/differentiation_utils/editable_module.py [hunk 11]
```diff
@@ -0,0 +1,502 @@
...
+    converted to pure functions for higher order derivatives purpose.
+    """
+
+    def getparams(self, methodname: str) -> Sequence[torch.Tensor]:
```
Comment (rbharath) [inline deepchem/utils/differentiation_utils/editable_module.py:24]: Add a docstring

File: deepchem/utils/differentiation_utils/editable_module.py [hunk 12]
```diff
@@ -0,0 +1,502 @@
...
+        # Returns a list of tensor parameters used in the object's operations
+
+        paramnames = self.cached_getparamnames(methodname)
+        return [get_attr(self, name) for name in paramnames]
+
+    def setparams(self, methodname: str, *params) -> int:
```
Comment (rbharath) [inline deepchem/utils/differentiation_utils/editable_module.py:30]: Add a docstring

File: deepchem/utils/differentiation_utils/editable_module.py [hunk 13]
```diff
@@ -0,0 +1,502 @@
...
+        methodname = method.__name__
+
+        # assert if the method preserve the float tensors of the object
+        self.__assert_method_preserve(method, *args, **kwargs)
+        self.__assert_get_correct_params(
+            method, *args,
+            **kwargs)  # check if getparams returns the correct tensors
+        print('"%s" method check done' % methodname)
+
+    def __assert_method_preserve(self, method, *args, **kwargs):
```
Comment (rbharath) [inline deepchem/utils/differentiation_utils/editable_module.py:248]: Add a docstring

File: deepchem/utils/differentiation_utils/editable_module.py [hunk 14]
```diff
@@ -0,0 +1,502 @@
...
+            if user_params_id[i] not in oper_params_id_set:
+                # if user_names[i] not in oper_names:
+                excess_names.append(user_names[i])
+        # if there are excess parameters, give warnings
+        if len(excess_names) > 0:
+            msg = "getparams for %s.%s has excess parameters: %s" % \
+                (clsname, methodname, ", ".join(excess_names))
+            warnings.warn(msg, stacklevel=3)
+
+    def __list_operating_params(self, method, *args, **kwargs):
```
Comment (rbharath) [inline deepchem/utils/differentiation_utils/editable_module.py:343]: Add a docstring

File: deepchem/utils/differentiation_utils/editable_module.py [hunk 15]
```diff
@@ -0,0 +1,502 @@
...
+import torch
+from typing import Sequence, Union, Dict, List
+from deepchem.utils.attribute_utils import get_attr, set_attr, del_attr
+
+__all__ = ["EditableModule"]
+
+torch_float_type = [torch.float32, torch.float, torch.float64, torch.float16]
+
+
+class EditableModule(object):
```
Comment (rbharath) [inline deepchem/utils/differentiation_utils/editable_module.py:18]: Add editable module to the docs

File: deepchem/utils/differentiation_utils/editable_module.py (hunk 10 above)
Comment (rbharath) [inline deepchem/utils/differentiation_utils/editable_module.py:20]: Add a paragraph explaining when we would want to use EditableModule

File: deepchem/utils/differentiation_utils/editable_module.py (hunk 15 above)
Comment (GreatRSingh) [reply deepchem/utils/differentiation_utils/editable_module.py:18]: Its added already, in the `utils.rst` file.
Do you mean that I should create another section for it.

**Pull request closed without merging**